Added default sender and subject values for email notification.

navya108 opened the pull request:
<!--
Filling out the template is required. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
Any pull request must pass the automated Travis tests.
-->

### Applicable Issues
https://github.com/eiffel-community/eiffel-intelligence/issues/401

### Description of the Change
Added sender and subject default values for email notification.

### Alternate Designs
<!-- Explain what other alternates were considered and why the proposed version was selected -->

### Benefits
We will always have email sender and subject defined, even if users do not defined them in their application.properties.

### Possible Drawbacks
<!-- What are the possible side-effects or negative impacts of the change? -->

### Sign-off
<!-- Sign the below certificate of origin, using your full name and e-mail address. -->
<!-- The certificate is copied from https://developercertificate.org/ -->

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.

Signed-off-by: <!-- <Your full name> <your-email@example.org> -->

e-pettersson-ericsson commented on src/main/java/com/ericsson/ei/notifications/EmailSender.java:
```diff
@@ -99,6 +99,11 @@ public MimeMessage prepareEmailMessage(String recipients, String mapNotification
      */
     private MimeMessage prepareEmail(String mapNotificationMessage, String emailSubject,
             String[] recipients) {
+        // Added default sender and subject for email notification
```
Perhaps extending the functionaltests for subscription notification with a new scenario to test this?

navya108 replied:
ok. I am implementing the unit and functional test cases for this functionality.

navya108 commented on src/main/java/com/ericsson/ei/notifications/EmailSender.java:
```diff
@@ -99,6 +99,11 @@ public MimeMessage prepareEmailMessage(String recipients, String mapNotification
             String[] recipients) {
+        // Added default sender and subject for email notification
+        if (sender.isEmpty() || subject.isEmpty()) {
+            sender = "noreply@ericsson.com";
```
In this case it is not possible because if we make the variable final we have to initiate it directly or in a constructor but we can't use the variables in if block.

e-pettersson-ericsson replied:
Ah I meant the value only, for example: private static final String SUBJECT_VALUE = "Email notification";

navya108 replied:
Yah , If that is the case we can use the value but we can't reassign it.
In our Scenario as I am assigning the value after checking the condition so it is not possible to put as static final.

navya108 replied:
yah Emelie I have just tried putting static final and without if condition, its working fine. I will update the change.

a-breid-ericsson commented on src/main/java/com/ericsson/ei/notifications/EmailSender.java:
```diff
@@ -99,6 +99,11 @@ public MimeMessage prepareEmailMessage(String recipients, String mapNotification
     private MimeMessage prepareEmail(String mapNotificationMessage, String emailSubject,
             String[] recipients) {
+        // Added default sender and subject for email notification
+        if (sender.isEmpty() || subject.isEmpty()) {
```
This if statement is not needed since the values will never be empty, they will always have a default set at the top in the @Value field. 

Also the if statement in current state will replace the sender with default if the sender is given by the user but subject is not.

a-breid-ericsson commented on src/functionaltests/java/com/ericsson/ei/notifications/trigger/SubscriptionNotificationSteps.java:
```diff
@@ -187,6 +190,14 @@ public void send_one_previous_event() throws Throwable {
+            sender = "noreply@ericsson.com";
+            subject = "Email subscription notification";
+        }
+    }
```
Please update step "Mail subscriptions were triggered" with the method name "mail_subscriptions_were_triggered" to also validate the subject in the email, currentluy it only makes an assert on the "From" field that is the sender.

You should also test and ensure that it is possible to give only sender and not subject and the given sender should still be the value given by the user, and the other way around, a subject is given by the user but not sender.

a-breid-ericsson commented on src/functionaltests/resources/features/subscriptionNotification.feature:
```diff
@@ -22,3 +22,14 @@ Feature: Test Subscription Trigger
     And Wait for EI to aggregate objects
     And Failed notification db should contain 2 objects
+
+  @AddedDefaultValuesSenderAndSubject
```
This test could be extended to test also when one of the options is given but not the other, thus the given option should not be replaced with default.

e-pettersson-ericsson replied:
I agree, a new sceneario in which we test that the configured properties are used and not default values ... I don't see any mention of which configuration is used in any of the features. The first scenario "SubscriptionNotificationOK" could be extended with an assert that the email notifications has the correct subject + sender (matching input configuration)

navya108 replied:
In the first scenario I have extended assert for subject even. so should I remove the new scenario?

e-pettersson-ericsson replied:
Ah my mistake, yes I see that you extended the first scenario - good! 

Just an improvement tip: the feature for the new scenario "Test default values mail notification" there is no mention of how the configuration affects these values (i.e. first scenario we provide email configuration so EI will use those for notifications, and in next scenario we provide no configuration so EI will use default values instead) 

I imagine something like this (with the first scenario being the existing one just extended:

**Scenario 1**
Given the REST API is up and running
And email configuration is provided
We create subscriptions 
etc ...
Notification email contains \<our expected\> values

**Scenario 2**
Given the REST API is up and running
And email configuration is NOT provided
We create subscriptions 
etc ...
Notification email contains default values

a-breid-ericsson commented on src/main/java/com/ericsson/ei/notifications/EmailSender.java:
```diff
@@ -48,12 +48,12 @@
-    @Value("${email.sender}")
-    private String sender;
+    @Value("${email.sender:noreply@ericsson.com}")
+    private static final String sender = "noreply@ericsson.com";
```
A "private static final" variable should always be named in uppercase. You could remove static or rename the variables.

rogpers-ericsson replied:
Can we have ericsson.com in here?

e-pettersson-ericsson replied:
Question is: what to have instead as a default sender ? 🤔  Any suggestions?

rogpers-ericsson replied:
In principle, I think this should be a configurable thing. To be strict, perhaps no default may be used?

tobiasake replied:
I think its possible/ok to leave the sender address field empty in a mail, -> defaulting to a empty string. Only issue can be that Mail lib maybe require a non-empty string on that parameter and throw an exception. You can maybe set it to no-reply@dummy.com or something.

e-pettersson-ericsson replied:
Yes I wrote this issue because I noticed the lib throws an exception if the sender is empty. That is why we want a default if this is NOT configured in application.properties. Admin users will have the option to configure the sender + subject (which they probably will) but in case they forget we have default values.

tobiasake replied:
Another approach is to write a Log message that Email is disabled when mail parameter is missing or failed to connect to SMTP server. And add Status info in Frontend Status button / info box that "Email is disabled" in backend.

tobiasake replied:
And leave the mail properties default empty and don't default anything. But when mail host property is specifued, some validation is required, e.g. senders mail address.

a-breid-ericsson commented on src/test/java/com/ericsson/ei/notifications/EmailSenderTest.java:
```diff
@@ -43,14 +44,29 @@
 
+    @Test
+    public void testDefaultValues() {
+    if (sender == null || subject == null) {
```
in this case if one value is given but not the other both will be set to be default not just the missed one.

e-pettersson-ericsson replied:
Isn't it enough to extend these unit tests with more scenarios, to test? 1) only subject was given, 2) only sender was given 3) no configuration was given - default should be added. 
Then we don't have to extend the cucumber scenarios with several scenarios, but just one 😄

Christoffer-Cortes replied:
Currently I don't understand what this tests. The values are fetched from the properties and they have a defined default. This means they cannot be null so the test would always fall to the else statement.

Christoffer-Cortes commented on src/test/java/com/ericsson/ei/notifications/EmailSenderTest.java:
```diff
@@ -43,14 +44,33 @@
     }
 
+    @Test
+    public void testDefaultValues() {
```
I think this test can be removed since it currently does not seem to be testing anything.

navya108 replied:
Emelie has asked for the test case, so I have added the change.

Christoffer-Cortes replied:
You could make an instance of EmailSender and then get SENDER and SUBJECT and make sure they are set.
Then you are testing the logic through the EmailSender class.
There is also no assertion in the test so it would accept any outcome. A test should include an assert that checks some expected outcome of the test, unless this for reason is not possible. In your case you instantiate the class, get the values and assert that they were set to the default values.
You can also check that you can also override the defaults as expected by setting the environment variables first and then make an EmailSender instance and check that SENDER and SUBJECT were indeed set to something other than default.

e-pettersson-ericsson replied:
After seeing the funcitonal test scenario added above, I agree that this test is not really needed. This is basically testing that Spring has assigned the correct values for the properties, and we want to test our logic - not Spring's. If the annotations are used correctly (see above comments by @a-breid-ericsson and @Christoffer-Cortes ) only Spring assigns the value to the variables and we don't have to do this in the EmailSender class. But we do want to check that if no configuration was given we use the default values - and this is tested in the functional tests.

Christoffer-Cortes commented on src/main/java/com/ericsson/ei/notifications/EmailSender.java:
```diff
@@ -48,17 +48,31 @@
     /**
+     * @return the sender
+     */
+    public String getSender() {
```
EI uses lombok @Getter so you don't beed to explicitly define these types of methods. You can read about it here: https://www.baeldung.com/lombok-ide
You may be getting an error currently and/or be able unable to get these varibles unless installing lombok.
Once you do you won't need these methods anymore.

navya108 replied:
okay

Christoffer-Cortes commented on src/main/java/com/ericsson/ei/notifications/EmailSender.java:
```diff
@@ -48,17 +48,31 @@
+    /**
+     * @return the subject
+     */
+    public String getSubject() {
```
See above.

e-pettersson-ericsson commented on src/functionaltests/java/com/ericsson/ei/notifications/trigger/SubscriptionNotificationSteps.java:
```diff
@@ -187,6 +195,14 @@ public void send_one_previous_event() throws Throwable {
 
+    @Then("^Default values mail notification are added$")
+    public void default_values_mail_notification() {
+        if (emailSender.getSender() != null && emailSender.getSubject() != null) {
```
This if statement is not needed, sin we always want to check the assert. If the value is null then the assertion goes wrong either way.

navya108 replied:
okay You mean this scenario is not required? shall I remove it? As I have also added the subject assert in mail_subscriptions_were_triggered method.

e-pettersson-ericsson replied:
No the scenario should still be there, but you can remove line 200: "if (emailSender.getSender() != null && emailSender.getSubject() != null)" and always check the assert on these values.

navya108 replied:
okay

a-breid-ericsson commented on src/main/java/com/ericsson/ei/notifications/EmailSender.java:
```diff
@@ -48,17 +48,31 @@
-    @Value("${email.sender}")
-    private String sender;
+    @Value("${email.sender:noreply@ericsson.com}")
+    private static final String SENDER = "noreply@ericsson.com";
```
The default value is now given 2 times, both as @Value and decleration. If Eclipse ot the IDE you are using is complaining about not declared variables then the @Value maybe will not work.

a-breid-ericsson replied:
A test must also ensure that a given value will not be set to default.

a-breid-ericsson replied:
After some googling I found that this will not work, a @Value from spring cannot change a final static variable, thus final static must be removed. And variable name changed to lower case :)

navya108 replied:
yes If final static is used default values cant be changed. If that is the case required  ( where irrespective of user inputs, you want sender and subject as noreply.ericsson.com and Email Subscription Notification respectively.) ...these changes are suitable.

But what I have observed is, if user gives email and subject in application.proporties . Those will not override the sender and subject values in EmailSender class as they are set to final static. So we have to check it putting ifconditions ( if email.sender is empty then assign the value ) as I have done previously of my First push.

so please let me know for which case the changes are required so that accordingly I can proceed.

Christoffer-Cortes replied:
The variable can't be static final then. Just type them out like this:
@Value("${email.sender:noreply@ericsson.com}")
private String sender;
That should make it possible for them to fetch the values from properties.
Also note that the sender variable does not need to be explicitly set, it is already set through the @Value annotation.

tobiasake replied:
Remove "ericsson" from the source code. change to some thing else or skip defaulting a value.
Let the Validation Properties discussion/issue solve it later.

a-breid-ericsson commented:
We had a small discussion about this issue in the team but not the entire team. We have thoughts if maybe the sollution should be to have validation of all the given input options in application.properties att application start, and in that case sender and subject must always be set (and all other required parameters in the application.properties). And if so then the issue for this will not be needed. We will discuss this further with Santosh and Raja, cause that may be a better long term sollution.

e-pettersson-ericsson commented:
@a-breid-ericsson, it sounds like you want to hold off completely on this PR and instead go for a new issue/PR in which we add validation of all configuration being properly set? While I do agree a proper validation of all mandatory configuration is needed, it will take time before it is merged into master, so why not have these changes in (once they are approved) while waiting for the other PR?

navya108 commented:
As per my observation and testing, If email.sender and email.subject are empty in application.proporties. Values what we are mentioning as below 
@value("${email.sender:noreply@ericsson.com}")
private String sender;
are not getting set as default values. And still sender and subject are reflecting as Empty values.

So I have added the conditions for checking isEmpty for both subject and sender and updated the test cases.
 
( **The below case** 
@value("${email.sender:noreply@ericsson.com}")
private String sender;
is working when entire fields i.e email.sender and email.subject are not present in application.proporties.)

Please check those changes and let me know if this approach is fine or not.

e-pettersson-ericsson commented on src/main/java/com/ericsson/ei/notifications/EmailSender.java:
```diff
@@ -83,13 +83,29 @@ public void sendEmail(MimeMessage message) throws NotificationFailureException {
+     * at application.proporties are empty.
+     * */
+    private void setDefaultValues() {
+        if (sender.isEmpty() && subject.isEmpty()) {
```
I think this can be simplified with only two if statements right? To first check sender and then check subject. A few less lines of code 😄

e-pettersson-ericsson commented on src/functionaltests/java/com/ericsson/ei/notifications/trigger/SubscriptionNotificationSteps.java:
```diff
@@ -195,16 +195,16 @@ public void send_one_previous_event() throws Throwable {
-        }
+    @Then("^Then Notification email contains <our expected> values")
+    public void notification_email_contains_expected_values() {
+        sender = "noreply@ericsson.com";
```
This doesn't contain any assert. You can send in the two String values from the feature file and then assert that both these values are set in the notification:

assertEquals("expected subject string", emailSender.getSubject())

e-pettersson-ericsson commented on src/functionaltests/resources/features/subscriptionNotification.feature:
```diff
@@ -8,6 +8,7 @@ Feature: Test Subscription Trigger
     And Subscriptions are created
     When I send Eiffel events
     And Wait for EI to aggregate objects
+    Then Notification email contains <our expected> values
```
Instead of \<our expected\> values I want you to explicitly write which values we are using and then they can be sent to the cucumber test from the feature file using REGEX. Check other feature files how this is done.

e-pettersson-ericsson commented on src/main/java/com/ericsson/ei/notifications/EmailSender.java:
```diff
@@ -96,12 +96,10 @@ public MimeMessage prepareEmailMessage(String recipients, String mapNotification
-            sender = "noreply@ericsson.com";
-            subject = "Email Subscription Notification";
-        } else if (!sender.isEmpty() && subject.isEmpty()) {
+       if (subject.isEmpty()) {
```
Much beter! I wasn't aware that Spring interpreted it as empty values when the property is declared but no value. But this approach is good.

tobiasake commented on src/main/java/com/ericsson/ei/notifications/EmailSender.java:
```diff
@@ -48,17 +48,17 @@
 
     @Getter
-    @Value("${email.sender}")
+    @Value("${email.sender:noreply@ericsson.com}")
```
I change "...ericsson.com" to something else, e.g. domain.com or something.

tobiasake commented on src/main/java/com/ericsson/ei/notifications/EmailSender.java:
```diff
@@ -90,6 +90,19 @@ public MimeMessage prepareEmailMessage(String recipients, String mapNotification
+     * */
+    private void setDefaultValuesEmailSenderSubject() {
+       if (subject.isEmpty()) {
+            subject = "Email Subscription Notification";
```
Is this if-statements really needed since it set default to  "Email Subscription Notification" on line 55, if subject string is empty.
So whole this if-statement can be removed, if understand the code correctly.

e-pettersson-ericsson replied:
Apparently Spring doesn't fill in default values if the property is present in application.properties but without values. It interprets "" as a value and will not override it ... So that's why this check is needed. Perhaps add this in the javadoc for the method to clarify future questions 😄

navya108 replied:
yah as per my testing I observed that even the values are given at @value in Emailsender class, those are not been assigned  when the user gives as below in application.proporties.
email.sender:
email.subject:

when I tried to print sender and subject.. still I can see those empty values.
so I have kept those if statments to handle isEmpty values for sender and subject.

This @Value("${email.sender:noreply@domain.com}") and @Value("${email.subject:Email Subscription Notification}") line is working only when the entire field email.sender and email.subject are not present in application.proporties.

e-pettersson-ericsson replied:
I saw the javadoc was updated in a previous commit, nice!

tobiasake replied:
Then I guess default value in @Value ":noreply@....." is not needed , since its defaulted here instead.

navya108 replied:
Right now its not needed , I just kept it considering the situation (if in case the fields email.sender and email.subject) are not present in application.proporties

e-pettersson-ericsson commented on src/test/java/com/ericsson/ei/notifications/EmailSenderTest.java:
```diff
@@ -51,9 +51,9 @@ public void sendEmailThrowsException() throws Exception {
     public void testDefaultValues() {
-        assertEquals("noreply@ericsson.com", emailSender.getSender());
+        assertEquals("noreply@domain.com", emailSender.getSender());
         assertEquals("Email Subscription Notification", emailSender.getSubject());
```
Shouldn't this test call upon emailSender.prepareEmailMessage? Then we assert the values of subject and sender in the created message? If we send in a null value for this parameter in the test, we expect the default values to have been added. And I don't think there is a need for the expected NullPointerException then.

navya108 replied:
yah I have tried putting upon emailSender.prepareEmailMessage but still facing the error.
Actually in EmailSenderTest class its mentioned previously javadocs that 
 /*
     * prepareEmail is not possible to test due too much hidden Spring mumbo jumbo doing things in
     * background.
     */
so I think should remove the testcase.

e-pettersson-ericsson replied:
I took a look at the code again, and it seems very difficult to mock the MimeMessage and MimeMessageHelper... Since the default values doesn't get added until we actually use the prepareEmail method this test is rendered useless. I say remove this unit test and instead keep the cucumber functional test which actually sends an email with the correct subject + sender...

e-pettersson-ericsson replied:
I agree with you, remove this test case then.

e-pettersson-ericsson commented:
I see that Codacy webhooks are not working properly, but it complains about unused imports in your PR :) I'm looking into how to fix that Codacy checks are also enabled for PRs.

e-pettersson-ericsson commented on src/main/java/com/ericsson/ei/notifications/EmailSender.java:
```diff
@@ -90,6 +90,19 @@ public MimeMessage prepareEmailMessage(String recipients, String mapNotification
+       if (subject.isEmpty()) {
+            subject = "Email Subscription Notification";
+        }
+       if (sender.isEmpty()) {
```
Use StringUtils.isEmpty to avoid NullPointer in these if statements

navya108 replied:
okay

e-pettersson-ericsson commented on src/main/java/com/ericsson/ei/notifications/EmailSender.java:
```diff
@@ -95,10 +96,10 @@ public MimeMessage prepareEmailMessage(String recipients, String mapNotification
             subject = "Email Subscription Notification";
         }
-       if (sender.isEmpty()) {
+       if (StringUtils.isEmpty(sender) == true) {
```
No need for the "== true". The StringUtils.isEmpty(sender) gives a boolean result which can be used as is, in the if statement.

e-pettersson-ericsson commented on src/test/java/com/ericsson/ei/notifications/EmailSenderTest.java:
```diff
@@ -18,10 +18,12 @@
 import org.mockito.InjectMocks;
 import org.mockito.Mock;
 import org.mockito.junit.MockitoJUnitRunner;
+import org.springframework.beans.factory.annotation.Value;
```
It looks like there are unused imports in this file - remove them.

e-pettersson-ericsson replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused import - org.springframework.beans.factory.annotation.Value.](https://app.codacy.com/manual/eiffel-intelligence-maintainers/eiffel-intelligence/pullRequest?prid=4891534)

e-pettersson-ericsson replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid unused imports such as 'org.springframework.beans.factory.annotation.Value'](https://app.codacy.com/manual/eiffel-intelligence-maintainers/eiffel-intelligence/pullRequest?prid=4891534)

e-pettersson-ericsson commented on src/test/java/com/ericsson/ei/notifications/EmailSenderTest.java:
```diff
@@ -18,10 +18,12 @@
 import javax.mail.internet.MimeMessage;
 
 import org.junit.Test;
+import static org.junit.Assert.*;
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid unused imports such as 'org.junit.Assert'](https://app.codacy.com/manual/eiffel-intelligence-maintainers/eiffel-intelligence/pullRequest?prid=4891534)

tobiasake commented on src/main/java/com/ericsson/ei/notifications/EmailSender.java:
```diff
@@ -48,17 +49,17 @@
 
     @Getter
-    @Value("${email.subject}")
+    @Value("${email.subject:Email Subscription Notification}")
```
Is default subject really needed here, since its defaulted later in the code.?
